src/Kestrel.Core/HttpsConnectionAdapterOptions.cs

```diff
@@ -29,14 +29,25 @@ public HttpsConnectionAdapterOptions()
 
         /// <summary>
         /// <para>
-        /// Specifies the server certificate used to authenticate HTTPS connections.
+        /// Specifies the server certificate used to authenticate HTTPS connections. This is ignored if ServerCertificateSelector is set.
         /// </para>
         /// <para>
         /// If the server certificate has an Extended Key Usage extension, the usages must include Server Authentication (OID 1.3.6.1.5.5.7.3.1).
         /// </para>
         /// </summary>
         public X509Certificate2 ServerCertificate { get; set; }
 
+        /// <summary>
+        /// <para>
+        /// A callback that will be invoked to dynamically select a server certificate. This is higher priority than ServerCertificate.
+        /// If SNI is not avialable then the name parameter will be null.
+        /// </para>
+        /// <para>
+        /// If the server certificate has an Extended Key Usage extension, the usages must include Server Authentication (OID 1.3.6.1.5.5.7.3.1).
+        /// </para>
+        /// </summary>
+        public Func<string, X509Certificate2> ServerCertificateSelector { get; set; }
```

Comment:
Consider passing through a state object, either a Kestrel specific object or the SslStream itself. That's what SslStream passes to their callback (as an Object).

Reply:
So more of a "sender" than a state object? I don't think a state object is necessary since any closures would be allocated only once for the lifetime of the server.

Reply:
Yes, a sender. Is there any connection specific state we think would be interesting to expose to this callback?

Reply:
Why not just send the SslStream any properties are likely to be exposed on that in the future and means you won't have to change the kestrel side for every change.

Reply:
SslStream is largely uninitialized at this stage in the handshake. See https://github.com/dotnet/corefx/pull/28278#discussion_r177871697

Reply:
I agree that is the way it is today. But as we have client hello parsing started it's just a backdoor to other parts being parsed and populated ;) (don't tell anyone) and if these features are populated then there will be no lightup required in kestrel.

Reply:
Maybe the ConnectionContext would be useful.

Reply:
Updated to pass through the sender. Trying to pass though per-connection state requires allocating a per-connection closure. We'll follow up in RC to see if that makes sense.

Reply:
Not that we have to do this now, but a per-connection closure allocation isn't a big deal in the context of initializing an HTTPS connection in Kestrel right now.

Reply:
Yeah, I realized that looking at the client cert validation code. I added it back for now.